odoo/addons/base/models/ir_model.py

```diff
@@ -874,6 +874,48 @@ def _drop_column(self):
 
         return True
 
+    def _get_inherit_to_remove(self, modules_to_remove):
+        modules_to_remove = set(modules_to_remove)
+        inherit_fields_to_remove_ids = OrderedSet()
+        model_name_to_field_records = self.grouped('model')
+
+        for model_name, field_records in model_name_to_field_records.items():
+            if model_name not in self.env:
+                continue
+
+            models_to_check = OrderedSet()
+
+            inherit_children_models = list(self.pool[model_name]._inherit_children)
+            while inherit_children_models:
+                current_model = inherit_children_models.pop()
+                models_to_check.add(current_model)
+                inherit_children_models.extend(m for m in self.pool[current_model]._inherit_children if m not in models_to_check)
+
+            for model_name_ in models_to_check:
+                model_class = self.pool[model_name_]
+                for field_record in field_records:
+                    field = model_class._fields.get(field_record.name)
+                    if field and all(m in modules_to_remove for m in field._modules):
+                        inherit_fields_to_remove_ids.add(self._get(model_name_, field_record.name).id)
+
+        inherit_fields_to_remove = self.browse(inherit_fields_to_remove_ids) - self
+        if inherit_fields_to_remove:
+            # when upgrade some customzied modules may not be loaded which may also override the fields
+            self.env['ir.model.data'].flush_model()
+            self.env['ir.module.module'].flush_model()
+            fields_not_loaded_ids = self.env.execute_query(SQL("""
+                SELECT array_agg(imd.res_id)
+                FROM ir_model_data imd
+                JOIN ir_module_module imm
+                ON imd.module NOT IN %s
+                AND imm.state IN ('to upgrade', 'installed')
+                AND imd.res_id IN %s
+                AND imd.model = 'ir.model.fields'
```

Comment:
I think something is missing in the JOIN. I'd also suggest moving what looks like filtering (not necessary for the join) to the WHERE.
```suggestion
                FROM ir_model_data imd
                JOIN ir_module_module imm
                  ON imd.module = imm.name
               WHERE imd.module NOT IN %s
                 AND imm.state IN ('to upgrade', 'installed')
                 AND imd.res_id IN %s
                 AND imd.model = 'ir.model.fields'
```
I'm still going over the PR, leaving this comment because I think something is missing here...

Reply:
Yes the condition `imd.module = imm.name` was missing
but I think `ir_module_data` is not a small table, it is better to `JOIN` with `AND imd.res_id IN %s` since the number of checked fields is really limited.

Reply:
I think it doesn't change anything from the query planner point of view. At least in a noticeable way given the heavier processing being done here. From a simple test, the query plan on a random runbot DB is the same: 22K entries in imd, 978 modules installed, the query runs in 10ms.